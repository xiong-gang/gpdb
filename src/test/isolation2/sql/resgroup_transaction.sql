-- ----------------------------------------------------------------------
-- Test: manage resource group in transaction
-- ----------------------------------------------------------------------

--start_ignore
DROP RESOURCE GROUP rg_test_group;
--end_ignore

-- helper view to check the resgroup status
CREATE OR REPLACE VIEW rg_test_monitor AS
	SELECT groupname, concurrency, proposed_concurrency, cpu_rate_limit
	FROM gp_toolkit.gp_resgroup_config
	WHERE groupname='rg_test_group';

-- ----------------------------------------------------------------------
-- Test: new resource group created in transaction then rollback
-- ----------------------------------------------------------------------

-- CREATE then ROLLBACK
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- CREATE, DROP then ROLLBACK
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;

	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- CREATE, ALTER then ROLLBACK
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- CREATE, ALTER, DROP then ROLLBACK
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;

	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- ----------------------------------------------------------------------
-- Test: new resource group created in transaction then commit
-- ----------------------------------------------------------------------

-- CREATE then COMMIT
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;
DROP RESOURCE GROUP rg_test_group;
SELECT * FROM rg_test_monitor;

-- CREATE, DROP then COMMIT
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;

	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;

-- CREATE, ALTER then COMMIT
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;
DROP RESOURCE GROUP rg_test_group;
SELECT * FROM rg_test_monitor;

-- CREATE, ALTER, DROP then COMMIT
BEGIN;
	CREATE RESOURCE GROUP rg_test_group
		WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;

	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;

-- ----------------------------------------------------------------------
-- Test: manage existing resource group in transaction then rollback
-- ----------------------------------------------------------------------

CREATE RESOURCE GROUP rg_test_group
	WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);

-- DROP then ROLLBACK
BEGIN;
	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- ALTER then ROLLBACK
BEGIN;
	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- ALTER, DROP then ROLLBACK
BEGIN;
	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;

	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- ----------------------------------------------------------------------
-- Test: manage existing resource group in transaction then commit
-- ----------------------------------------------------------------------

-- DROP then COMMIT
BEGIN;
	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;
CREATE RESOURCE GROUP rg_test_group
	WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
SELECT * FROM rg_test_monitor;

-- ALTER then COMMIT
BEGIN;
	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;

-- ALTER, DROP then COMMIT
BEGIN;
	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 12;
	SELECT * FROM rg_test_monitor;

	ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.12;
	SELECT * FROM rg_test_monitor;

	DROP RESOURCE GROUP rg_test_group;
	SELECT * FROM rg_test_monitor;
COMMIT;
SELECT * FROM rg_test_monitor;

-- ----------------------------------------------------------------------
-- Test: manage resource group in subtransaction must fail
-- ----------------------------------------------------------------------

-- CREATE in subtransaction
BEGIN;
	SAVEPOINT sub1;
		CREATE RESOURCE GROUP rg_test_group
			WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
ROLLBACK;
SELECT * FROM rg_test_monitor;

-- ALTER and DROP in subtransaction
CREATE RESOURCE GROUP rg_test_group
	WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
SELECT * FROM rg_test_monitor;
BEGIN;
	SAVEPOINT sub1;
	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ROLLBACK;
BEGIN;
	SAVEPOINT sub1;
	ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ROLLBACK;
BEGIN;
	SAVEPOINT sub1;
	DROP RESOURCE GROUP rg_test_group;
ROLLBACK;
DROP RESOURCE GROUP rg_test_group;
SELECT * FROM rg_test_monitor;

-- cleanup
DROP VIEW rg_test_monitor;
