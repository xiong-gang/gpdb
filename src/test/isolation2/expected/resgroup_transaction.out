-- ----------------------------------------------------------------------
-- Test: manage resource group in transaction
-- ----------------------------------------------------------------------

--start_ignore
DROP RESOURCE GROUP rg_test_group;
ERROR:  resource group "rg_test_group" does not exist
--end_ignore

-- helper view to check the resgroup status
CREATE OR REPLACE VIEW rg_test_monitor AS SELECT groupname, concurrency, proposed_concurrency, cpu_rate_limit FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';
CREATE

-- ----------------------------------------------------------------------
-- Test: new resource group created in transaction then rollback
-- ----------------------------------------------------------------------

-- CREATE then ROLLBACK
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- CREATE, DROP then ROLLBACK
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- CREATE, ALTER then ROLLBACK
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- CREATE, ALTER, DROP then ROLLBACK
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)

DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- ----------------------------------------------------------------------
-- Test: new resource group created in transaction then commit
-- ----------------------------------------------------------------------

-- CREATE then COMMIT
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)
DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- CREATE, DROP then COMMIT
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- CREATE, ALTER then COMMIT
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)
DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- CREATE, ALTER, DROP then COMMIT
BEGIN;
BEGIN
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)

DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- ----------------------------------------------------------------------
-- Test: manage existing resource group in transaction then rollback
-- ----------------------------------------------------------------------

CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE

-- DROP then ROLLBACK
BEGIN;
BEGIN
DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

-- ALTER then ROLLBACK
BEGIN;
BEGIN
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

-- ALTER, DROP then ROLLBACK
BEGIN;
BEGIN
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)

DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

-- ----------------------------------------------------------------------
-- Test: manage existing resource group in transaction then commit
-- ----------------------------------------------------------------------

-- DROP then COMMIT
BEGIN;
BEGIN
DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)

-- ALTER then COMMIT
BEGIN;
BEGIN
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.10          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.11;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|11         |11                  |0.11          
(1 row)

-- ALTER, DROP then COMMIT
BEGIN;
BEGIN
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 12;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|12         |12                  |0.11          
(1 row)

ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.12;
ALTER
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|12         |12                  |0.12          
(1 row)

DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)
COMMIT;
COMMIT
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- ----------------------------------------------------------------------
-- Test: manage resource group in subtransaction must fail
-- ----------------------------------------------------------------------

-- CREATE in subtransaction
BEGIN;
BEGIN
SAVEPOINT sub1;
SAVEPOINT
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
ERROR:  CREATE RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK;
ROLLBACK
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- ALTER and DROP in subtransaction
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=10, cpu_rate_limit=.10, memory_limit=.10);
CREATE
SELECT * FROM rg_test_monitor;
groupname    |concurrency|proposed_concurrency|cpu_rate_limit
-------------+-----------+--------------------+--------------
rg_test_group|10         |10                  |0.10          
(1 row)
BEGIN;
BEGIN
SAVEPOINT sub1;
SAVEPOINT
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ERROR:  ALTER RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK;
ROLLBACK
BEGIN;
BEGIN
SAVEPOINT sub1;
SAVEPOINT
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 11;
ERROR:  ALTER RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK;
ROLLBACK
BEGIN;
BEGIN
SAVEPOINT sub1;
SAVEPOINT
DROP RESOURCE GROUP rg_test_group;
ERROR:  DROP RESOURCE GROUP cannot run inside a subtransaction
ROLLBACK;
ROLLBACK
DROP RESOURCE GROUP rg_test_group;
DROP
SELECT * FROM rg_test_monitor;
groupname|concurrency|proposed_concurrency|cpu_rate_limit
---------+-----------+--------------------+--------------
(0 rows)

-- cleanup
DROP VIEW rg_test_monitor;
DROP
